CONTRIBUTING.md

```diff
@@ -0,0 +1,205 @@
+* Offer alternative implementations, but assume the author already considered
+  them. ("What do you think about using an off the shelf library here?")
+* Seek to understand the author's perspective.
+* Sign off on the pull request with a :thumbsup: or "Ready to merge" comment.
+
+## Quality Control for Contributed Code
+
+### Regression Tests
+
+The main regression tests for HClib are located in `test/c` and `test/cpp`.
+The `test_all.sh` scripts in each of those directories will automatically
+build and run all tests in the respective test suite.
+
+All new commits should pass all of the regression tests.
+When you open a pull request, Travis-CI will automatically run
+the full regression suite against your cumulative change
+(not against individual commits).
+Please run the tests locally against each commit to ensure
+that intermediate commits do not fail any tests
+(which breaks the ability to use `git bisect` and similar tools).
+
+If your pull request fixes a bug, it should also include a regression test for detecting that bug (with the appropriate changes in `test/c`, `test/cpp` and `test_all.sh`. This is to reduce the possibility of reintroducing the bug with some later changes.
+
+If your pull request implements a new feature, it should also include tests in `test/c` and `test/cpp` that test that feature.
+
+### Static Error Checking
+
+As part of the development workflow for HClib, any newly committed code
+should be checked using standard static checking tools.
+
+In particular, run
+[`cppcheck`](https://sourceforge.net/projects/cppcheck/)
+on all modified files.
+cppcheck should be run by `cd`-ing to `tools/cppcheck` and executing
+the `run.sh` script from there (this assumes `cppcheck` is on your path).
+Any new errors printed by `cppcheck` should be addressed before committing.
+
+### Code Formatting
+
+You should also run `astyle` on all modified or newly created files.
+`astyle` is a source code auto-formatter.
+Simply `cd` to `tools/astyle` and execute the `run.sh` script.
+This assumes you have `astyle` installed and it is on your path.
+
+## Summary
+
+Here are some useful tips for making good pull requests (borrowed and heavily modified from 
+[`Mark Seemann's blog`](http://blog.ploeh.dk/2015/01/15/10-tips-for-better-pull-requests/)
+):
+
+1. *Make it small.*
+A small, focused Pull Request gives you the best chance of having it accepted. The time it takes to review a Pull Request 
+is typically non-linear in the size of the request. Don't make your reviewers work.
+
+2. *Do only one thing.*
+Your Pull Request should address only a single concern. If it addresses multiple concerns (say A, B and C), the reviewers might 
+be fine with A and B, but have issues with C, delaying the merge of your Pull Request. This might unecessarily complicate the merge 
+of your perfectly fine A and B contributions if there were some updates to the repository in the meantime.
+
+3. *Watch your line width.*
```

Comment:
I would add that we encourage any newly merged code to be no wider than 80 columns.